### ISSUE TITLE
elan: does not keep interpreter alive through nix gc

### ISSUE THREAD

**Issue by nomeata** [NixOS/nixpkgs#323961]
The current `elan` package has a patch at
[0001-dynamically-patchelf-binaries.patch](https://github.com/NixOS/nixpkgs/blob/0f56e3221392452f2c38e3ddf2eba03abda6bf47/pkgs/applications/science/logic/elan/0001-dynamically-patchelf-binaries.patch) that `patchelf`s the downloaded binary with the current interpreter. This is great, and works well, it seems.

One wrinkle is that (unless I am missing something) nothing prevents `nix gc` from freeing that interpreter, e.g. after a system update, and thus breaking the patched `lean` binary.

Is there a reasonably simple way to prevent that?

For example, adding an otherwise unused symlink to the interpreter in `~/.elan/toolchains/leanprover--lean4---v4.10.0-rc1/elf-interpreter` that itself is then linked to `/nix/var/nix/gcroots/auto` from. Would that keep it alive exactly as long as that directory in `~/.elan/toolchains/` exists?

Pinging @kha and @gebner as people who have touched that patch before.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by Kha** [comment]
Yes, we were aware of that limitation. Your suggested fix sounds reasonable to me.

**Comment by nomeata** [comment]
Ok, maybe I’ll try that at some point, unless someone beats me to it.
